File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-discover/src/state/discoverState.ts
import type { DiscoverPageResult } from "../types/discoverModels.js";

export interface DiscoverState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: DiscoverPageResult<unknown>;
  readonly error?: string;
}

export function initialDiscoverState(): DiscoverState {
  return { status: "idle" };
}

export function reduceDiscoverState(
  state: DiscoverState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: DiscoverPageResult<unknown> }
    | { type: "failed"; error: string },
): DiscoverState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_discover/lib/src/state/discover_state.dart
import '../models/discover_models.dart';

enum DiscoverStatus { idle, loading, ready, error }

class DiscoverState {
  const DiscoverState({
    this.status = DiscoverStatus.idle,
    this.lastResult,
    this.error,
  });

  final DiscoverStatus status;
  final DiscoverPageResult<Object?>? lastResult;
  final String? error;

  DiscoverState loading() => const DiscoverState(status: DiscoverStatus.loading);

  DiscoverState loaded(DiscoverPageResult<Object?> result) =>
      DiscoverState(status: DiscoverStatus.ready, lastResult: result);

  DiscoverState failed(String message) =>
      DiscoverState(status: DiscoverStatus.error, error: message);
}
